recurring_contract/models/move.py

```diff
@@ -178,8 +177,101 @@ def find_sum(numbers, target, partial=None):
+        the invoice lines, the due date, the payment reference, and the payment mode
+
+        :param inv_lines: a list of dictionaries containing the invoice lines to be added to the invoice, defaults to False
+        (optional)
+        :param due_date: The date at which the invoice should be paid, defaults to False (optional)
+        :param ref: The reference of the invoice, defaults to False (optional)
+        :param pay_mode_id: The payment mode to be used for the invoice, defaults to False (optional)
+        :return: A dictionary with the invoice_line_ids, date, payment_reference, and payment_mode_id.
+        """
+        # Ensure that the function receive only one invoice
+        self.ensure_one()
+        # Build the dictionnary
+        inv_val_dict = dict()
+        inv_val_dict[self.name] = {
+            'invoice_line_ids': self._build_invoice_lines(inv_lines) if inv_lines else self.invoice_line_ids.ids,
+            'date': due_date if due_date else self.date,
+            'payment_reference': ref if ref else self.payment_reference,
+            'payment_mode_id': pay_mode_id if pay_mode_id else self.payment_mode_id.id
+        }
+        return inv_val_dict
+
+    def _build_invoice_lines(self, inv_lines):
+        """
+        It creates a list of tuples that will be used to create, modify or delete invoice lines.
+
+        :param inv_lines: a dictionary of the form {product_id: {'amt': amount, 'contract': contract_id}}
+        :return: A list of tuples.
+        """
+        res = list()
+        # create or modify existing line
+        for product, dict_data in inv_lines.items():
+            amt = dict_data.get("amt")
+            if amt == 0:
+                res.append("to_cancel")
+            else:
+                line_ids = self.invoice_line_ids.filtered(lambda l: l.product_id == product).ids
+                if not line_ids:
+                    contract = dict_data.get("contract")
+                    contract_line = contract.contract_line_ids.filtered(lambda l: l.product_id == product)
+                    res.append((0, 0,
+                                {
+                                    "price_unit": amt,
+                                    "product_id": product.id,
+                                    'name': product.name,
+                                    'quantity': contract_line.quantity,
+                                    'contract_id': contract_line.contract_id.id,
+                                    'account_id': product.with_company(
+                                        contract_line.contract_id.company_id.id).property_account_income_id.id or False
+                                }
+                                ))
+                else:
+                    for line_id in line_ids:
+                        res.append((1, line_id, {"price_unit": amt}))
+
+        # Delete line if the product shouldn't be invoiced anymore
+        if len(inv_lines) < len(self.invoice_line_ids):
```

Comment:
Maybe the user changed a product in the contract lines but didn't remove a line. In that case this test will fail to remove the old line with the old product, and will create just a new line. You should rather compare the set of products from the invoice_lines with the set of products from the contract_lines to make sure you remove all of those that are no longer in the contract.